From 415d74377b612361f4122a95eca2649050e4fb50 Mon Sep 17 00:00:00 2001
From: Rohit Biradar <rohibira@qti.qualcomm.com>
Date: Fri, 11 Sep 2026 12:09:08 +0530
Subject: [PATCH] Add SELinux policy for camera-nhx

The proprietary camera testing script 'camera-nhx' launches
'nativehaltest' for camera validation. Define a dedicated SELinux policy
module, qcom_nhx, to transition the launcher and nativehaltest into
confined domains.

- qcom_nhx_launcher_t: domain for 'camera-nhx', including SoC detection
  through /sys/devices/soc0/soc_id.
- qcom_nhx_t: domain for 'nativehaltest', including access to the CamX
  framework, DMA, FastRPC, V4L video nodes, camera cache directories,
  and related camera runtime resources.

The 'camera-nhx' launcher is installed at /usr/bin/camera-nhx.
'nativehaltest' is installed at
/usr/libexec/camx-<platform>/nativehaltest.

Upstream-Status: Inappropriate [Qualcomm specific change]

Signed-off-by: Rohit Biradar <rohibira@qti.qualcomm.com>
---
 policy/modules.conf                 |  6 ++
 policy/modules/services/qcom_nhx.fc | 10 ++++
 policy/modules/services/qcom_nhx.if | 89 ++++++++++++++++++++++++++++
 policy/modules/services/qcom_nhx.te | 91 +++++++++++++++++++++++++++++
 4 files changed, 196 insertions(+)
 create mode 100644 policy/modules.conf
 create mode 100644 policy/modules/services/qcom_nhx.fc
 create mode 100644 policy/modules/services/qcom_nhx.if
 create mode 100644 policy/modules/services/qcom_nhx.te

diff --git a/policy/modules.conf b/policy/modules.conf
new file mode 100644
index 0000000..f2dda81
--- /dev/null
+++ b/policy/modules.conf
@@ -0,0 +1,6 @@
+# Layer: services
+# Module: qcom_nhx
+#
+# Qualcomm NHX camera test launcher
+#
+qcom_nhx = module
diff --git a/policy/modules/services/qcom_nhx.fc b/policy/modules/services/qcom_nhx.fc
new file mode 100644
index 0000000..73c0a5a
--- /dev/null
+++ b/policy/modules/services/qcom_nhx.fc
@@ -0,0 +1,10 @@
+# File contexts for Qualcomm NHX camera test launcher
+
+# Label the camera-nhx launcher script
+/usr/bin/camera-nhx    --    gen_context(system_u:object_r:qcom_nhx_launcher_exec_t,s0)
+
+# Label nativehaltest binaries for different SoC families
+/usr/libexec/camx-[^/]*/nativehaltest    --    gen_context(system_u:object_r:qcom_nhx_exec_t,s0)
+
+# Label camera cache files
+/var/cache/camera(/.*)?    gen_context(system_u:object_r:qcom_camera_cache_t,s0)
diff --git a/policy/modules/services/qcom_nhx.if b/policy/modules/services/qcom_nhx.if
new file mode 100644
index 0000000..0e491a0
--- /dev/null
+++ b/policy/modules/services/qcom_nhx.if
@@ -0,0 +1,89 @@
+## <summary>Qualcomm NHX camera test launcher</summary>
+
+########################################
+## <summary>
+##	Execute camera-nhx launcher in the nhx_launcher domain.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed to transition.
+##	</summary>
+## </param>
+#
+interface(`qcom_nhx_launcher_domtrans',`
+	gen_require(`
+		type qcom_nhx_launcher_t, qcom_nhx_launcher_exec_t;
+	')
+
+	corecmd_search_bin($1)
+	domtrans_pattern($1, qcom_nhx_launcher_exec_t, qcom_nhx_launcher_t)
+')
+
+########################################
+## <summary>
+##	Execute camera-nhx launcher in the nhx_launcher domain, and
+##	allow the specified role the nhx_launcher domain.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed to transition.
+##	</summary>
+## </param>
+## <param name="role">
+##	<summary>
+##	Role allowed access.
+##	</summary>
+## </param>
+#
+interface(`qcom_nhx_launcher_run',`
+	gen_require(`
+		type qcom_nhx_launcher_t;
+	')
+
+	qcom_nhx_launcher_domtrans($1)
+	role $2 types qcom_nhx_launcher_t;
+')
+
+########################################
+## <summary>
+##	Execute nativehaltest in the nhx domain.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed to transition.
+##	</summary>
+## </param>
+#
+interface(`qcom_nhx_domtrans',`
+	gen_require(`
+		type qcom_nhx_t, qcom_nhx_exec_t;
+	')
+
+	corecmd_search_bin($1)
+	domtrans_pattern($1, qcom_nhx_exec_t, qcom_nhx_t)
+')
+
+########################################
+## <summary>
+##	Execute nativehaltest in the nhx domain, and
+##	allow the specified role the nhx domain.
+## </summary>
+## <param name="domain">
+##	<summary>
+##	Domain allowed to transition.
+##	</summary>
+## </param>
+## <param name="role">
+##	<summary>
+##	Role allowed access.
+##	</summary>
+## </param>
+#
+interface(`qcom_nhx_run',`
+	gen_require(`
+		type qcom_nhx_t;
+	')
+
+	qcom_nhx_domtrans($1)
+	role $2 types qcom_nhx_t;
+')
diff --git a/policy/modules/services/qcom_nhx.te b/policy/modules/services/qcom_nhx.te
new file mode 100644
index 0000000..2815ce2
--- /dev/null
+++ b/policy/modules/services/qcom_nhx.te
@@ -0,0 +1,91 @@
+policy_module(qcom_nhx, 1.0)
+
+########################################
+#
+# Declarations
+#
+
+type qcom_nhx_launcher_t;
+type qcom_nhx_launcher_exec_t;
+init_daemon_domain(qcom_nhx_launcher_t, qcom_nhx_launcher_exec_t)
+
+type qcom_nhx_t;
+type qcom_nhx_exec_t;
+init_daemon_domain(qcom_nhx_t, qcom_nhx_exec_t)
+
+type qcom_camera_cache_t;
+files_type(qcom_camera_cache_t)
+
+########################################
+#
+# qcom_nhx_launcher local policy (camera-nhx)
+#
+
+gen_require(`
+	type initrc_t, etc_runtime_t, v4l_device_t, usr_t;
+')
+
+dev_read_sysfs(qcom_nhx_launcher_t)
+files_read_etc_files(qcom_nhx_launcher_t)
+corecmd_exec_shell(qcom_nhx_launcher_t)
+corecmd_exec_bin(qcom_nhx_launcher_t)
+libs_use_ld_so(qcom_nhx_launcher_t)
+libs_use_shared_libs(qcom_nhx_launcher_t)
+allow qcom_nhx_launcher_t qcom_nhx_launcher_t:fifo_file { read write getattr ioctl };
+allow qcom_nhx_launcher_t usr_t:lnk_file read;
+domtrans_pattern(qcom_nhx_launcher_t, qcom_nhx_exec_t, qcom_nhx_t)
+allow qcom_nhx_launcher_t initrc_t:unix_stream_socket { read write ioctl getattr };
+allow qcom_nhx_launcher_t etc_runtime_t:file getattr;
+allow qcom_nhx_launcher_t v4l_device_t:chr_file getattr;
+
+########################################
+#
+# qcom_nhx local policy (nativehaltest)
+#
+
+gen_require(`
+	type dma_device_t, fastrpc_device_t, fastrpc_secure_device_t, v4l_device_t;
+	type var_t, debugfs_t, fs_t, default_t;
+	type etc_runtime_t, usr_t, node_t, unreserved_port_t, tracefs_t;
+	type sysctl_vm_overcommit_t, sysctl_vm_t;
+')
+
+dev_read_sysfs(qcom_nhx_t)
+allow qcom_nhx_t qcom_nhx_exec_t:file { execute execute_no_trans };
+files_read_etc_files(qcom_nhx_t)
+corecmd_exec_shell(qcom_nhx_t)
+corecmd_exec_bin(qcom_nhx_t)
+libs_use_ld_so(qcom_nhx_t)
+libs_use_shared_libs(qcom_nhx_t)
+allow qcom_nhx_t qcom_nhx_t:fifo_file { read write getattr };
+allow qcom_nhx_t qcom_camera_cache_t:file { read write open getattr create append };
+allow qcom_nhx_t qcom_camera_cache_t:dir { search read write add_name open getattr create };
+filetrans_pattern(qcom_nhx_t, var_t, qcom_camera_cache_t, dir, "camera")
+filetrans_pattern(qcom_nhx_t, qcom_camera_cache_t, qcom_camera_cache_t, file)
+files_search_var(qcom_nhx_t)
+allow qcom_nhx_t qcom_nhx_t:unix_dgram_socket { create connect write };
+logging_send_syslog_msg(qcom_nhx_t)
+allow qcom_nhx_t dma_device_t:chr_file { read write open ioctl getattr };
+allow qcom_nhx_t fastrpc_device_t:chr_file { read write open ioctl getattr };
+allow qcom_nhx_t fastrpc_secure_device_t:chr_file { read write open ioctl getattr };
+allow qcom_nhx_t v4l_device_t:chr_file { read write open ioctl getattr };
+allow qcom_nhx_t qcom_nhx_t:capability sys_nice;
+allow qcom_nhx_t qcom_nhx_t:process signal;
+allow qcom_nhx_t initrc_t:unix_stream_socket { read write ioctl getattr };
+allow qcom_nhx_t fs_t:filesystem getattr;
+allow qcom_nhx_t default_t:dir { search read open };
+allow qcom_nhx_t debugfs_t:dir search;
+allow qcom_nhx_t self:process setsched;
+allow qcom_nhx_t etc_runtime_t:file { open read };
+allow qcom_nhx_t default_t:file { getattr open write };
+allow qcom_nhx_t node_t:tcp_socket node_bind;
+allow qcom_nhx_t self:qipcrtr_socket { create getattr read setopt write };
+allow qcom_nhx_t self:tcp_socket { bind create listen };
+allow qcom_nhx_t sysctl_vm_overcommit_t:file { open read };
+allow qcom_nhx_t sysctl_vm_t:dir search;
+allow qcom_nhx_t tracefs_t:dir search;
+allow qcom_nhx_t tracefs_t:file { open write };
+allow qcom_nhx_t unreserved_port_t:tcp_socket name_bind;
+allow qcom_nhx_t usr_t:dir { read watch };
+allow qcom_nhx_t usr_t:file { getattr open read };
+allow qcom_nhx_t usr_t:lnk_file read;
-- 
2.34.1

